modules/mobile/app/models/mobile/v0/adapters/vaos_v2_appointment.rb

```diff
@@ -0,0 +1,424 @@
+          # and optional extension (until the end of the string) (?:\sx(\d*))?$
+          phone_captures = phone&.match(/^\(?(\d{3})\)?.?(\d{3})-?(\d{4})(?:\sx(\d*))?$/)
+
+          if phone_captures.nil?
+            Rails.logger.warn(
+              'mobile appointments failed to parse VAOS V2 phone number',
+              phone: phone
+            )
+            return { area_code: nil, number: nil, extension: nil }
+          end
+
+          {
+            area_code: phone_captures[1].presence,
+            number: "#{phone_captures[2].presence}-#{phone_captures[3].presence}",
+            extension: phone_captures[4].presence
+          }
+        end
+
+        def healthcare_service
+          if va_appointment?
+            appointment[:service_name] || appointment[:physical_location]
+          else
+            appointment.dig(:extension, :cc_location, :practice_name)
+          end
+        end
+
+        def minutes_duration(minutes_duration)
+          # not in raw data, matches va.gov default for cc appointments
+          return 60 if appointment_type == APPOINTMENT_TYPES[:cc] && minutes_duration.nil?
+
+          minutes_duration
+        end
+
+        def va_appointment?
+          [APPOINTMENT_TYPES[:va],
+           APPOINTMENT_TYPES[:va_video_connect_gfe],
+           APPOINTMENT_TYPES[:va_video_connect_atlas],
+           APPOINTMENT_TYPES[:va_video_connect_home]].include?(appointment_type)
+        end
+
+        def comment
+          return embedded_data[:comment] if reason_code_contains_embedded_data?
+
+          appointment[:comment] || appointment.dig(:reason_code, :text)
+        end
+
+        def reason
+          return embedded_data[:reason_code] if reason_code_contains_embedded_data?
+
+          appointment.dig(:reason_code, :coding, 0, :code)
+        end
+
+        def patient_email
+          return embedded_data[:email] if reason_code_contains_embedded_data?
+
+          contact(appointment.dig(:contact, :telecom), CONTACT_TYPE[:email])
+        end
+
+        # the upstream server that hosts VA appointment requests (acheron) does not support some fields
+        # so the front end puts all of that data into a comment, which is returned from upstream
+        # as reason_code text. We must parse out some parts of that data. If any of those values is
+        # present, we assume it's an acheron appointment and use only acheron values for relevant attributes.
+        def reason_code_contains_embedded_data?
+          @reason_code_contains_embedded_data ||= embedded_data.values.any?
+        end
+
+        def embedded_data
+          @embedded_data ||= {
+            phone: embedded_data_matches&.[](2)&.strip,
+            email: embedded_data_matches&.[](3)&.strip,
+            preferred_dates: embedded_data_matches&.[](4)&.strip,
+            reason_code: embedded_data_matches&.[](5)&.strip,
+            comment: embedded_data_matches&.[](6)&.strip
```

Comment:
whats the likelihood that one of these fields is omitted when there isn't a value for it and throws off the index count?

Reply:
That's a good question. I considered this and had mixed feelings about it. The front end code does it all or nothing, like so:
```
    appointmentInfo = `${modality}|${phone}|${email}|${preferredDates}|${reasonCode}`;
    const reasonCodeBody = {
    text:
      data.reasonAdditionalInfo && isAcheron
        ? `${appointmentInfo}|${reasonText}`
        : null,
  };
```
So it appears to be all or nothing. 

I initially coded the parsing more cautiously, like so:
```
        def reason_code_contains_embedded_data?
          @reason_code_contains_embedded_data ||= embedded_data.values.any?
        end

        def embedded_data
          @embedded_data ||= begin
            comment = appointment.dig(:reason_code, :text)&.split('|')
            {
              modality: find_attribute(comment, 'preferred modality:'),
              phone: find_attribute(comment, 'phone number:'),
              email: find_attribute(comment, 'email:'),
              preferred_dates: find_attribute(comment, 'preferred dates:'),
              reason_code: find_attribute(comment, 'reason code:'),
              comment: find_attribute(comment, 'comments:')
            }
          end
        end

        def find_attribute(comment, attribute)
          return nil unless comment

          match = comment.find { |c| c.start_with?(attribute) }
          return nil unless match

          value = match.delete_prefix(attribute)
          return nil unless value

          value.strip
        end
```

But that is also somewhat fragile because it would break if anything (like a comment) contains a pipe. That's a bit of an edge case and is probably only something QA would do if they knew how this works under the hood. I was definitely on the fence about this one, so LMK if anyone has ideas. Given that users have the ability to add comments that are included in this field, there may not be a perfect way, so I think we just have to pick the method we think is most robust.

Reply:
ah got it. I'd say if the web front end is making that same assumption then we should be safe to as well.